### ISSUE TITLE
[debian] pyqt5 should take precedence over pyside

### ISSUE THREAD

**Issue by picca** [silx-kit/silx#1035]
Hello, I am working on the jessie-backports of the Debian packaging.

during the jessie build, pyqt5 AND pyside are installed.

Then I end up wit this error

```
INFO:silx.gui.qt._qt:PySide.QtOpenGL not available
Traceback (most recent call last):
  File "run_tests.py", line 369, in <module>
    unittest.defaultTestLoader.loadTestsFromNames(options.test_name))
  File "/usr/lib/python2.7/unittest/loader.py", line 130, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/lib/python2.7/unittest/loader.py", line 115, in loadTestsFromName
    test = obj()
  File "/<<BUILDDIR>>/silx-0.5.0+dfsg/.pybuild/pythonX.Y_2.7/build/silx/test/__init__.py", line 49, in suite
    from . import test_sx
  File "/<<BUILDDIR>>/silx-0.5.0+dfsg/.pybuild/pythonX.Y_2.7/build/silx/test/test_sx.py", line 73, in <module>
    from silx.gui import qt
  File "/<<BUILDDIR>>/silx-0.5.0+dfsg/.pybuild/pythonX.Y_2.7/build/silx/gui/qt/__init__.py", line 54, in <module>
    from ._qt import *  # noqa
  File "/<<BUILDDIR>>/silx-0.5.0+dfsg/.pybuild/pythonX.Y_2.7/build/silx/gui/qt/_qt.py", line 164, in <module>
    from ._pyside_dynamic import loadUi  # noqa
  File "/<<BUILDDIR>>/silx-0.5.0+dfsg/.pybuild/pythonX.Y_2.7/build/silx/gui/qt/_pyside_dynamic.py", line 44, in <module>
    from PySide.QtUiTools import QUiLoader
ImportError: No module named QtUiTools
E: pybuild pybuild:256: test: plugin custom failed with: exit code=1: env PYTHONPATH=/<<BUILDDIR>>/silx-0.5.0+dfsg/.pybuild/pythonX.Y_2.7/build GPU=False SILX_OPENCL=False SILX_TEST_LAW_MEM=True xvfb-run -a --server-args="-screen 0 1024x768x24" python2.7 run_tests.py -v
```

Will it be possible for silx to prefer pyqt5 over pyside.

**Comment by vallsv** [comment]
Hi. The algorithm should be the same as the one from matplotlib. @t20100 should be able to answer to you when he came back.

But for your specific issue, you can install a `python2-pyside.qtuitools` module, maybe we have to add it to our debian config files.

**Comment by picca** [comment]
Hello Valentin,

I can not add python-pyside.qtuitools since it is not part of testing AND because of this

https://lists.debian.org/debian-devel-announce/2017/08/msg00006.html

So for now I changed the logic... in order to have somethying working out of the box during the build.
(jessie, stretch, unstable)

Tell me if this is plain wrong :)

**Comment by t20100** [comment]
Hi,

@picca, the Qt binding import order is so that it is compatible with IPython's `%pylab qt` in jessie where for example the following fails:
```
import PyQt5.QtCore
%pylab qt
```

I would also prefer to have PyQt5 as the default (and supports PySide2 when it is ready), but that would break IPython usage in jessie...

**Comment by vallsv** [comment]
BTW maybe we can force using PyQt5 for our applications. The code is not the same. Any idea if we have a preferred target at the ESRF?

**Comment by t20100** [comment]
Good idea for the application to prefer PyQt5, but here the error is during testing.

Provided that Qt4 is no more maintained and will apparently be dropped by Debian as soon as they can, our main target should be PyQt5...

**Comment by t20100** [comment]
@picca, BTW you can provide this option `--qt-binding PyQt5` to `run_tests.py` to select the Qt binding to use.

**Comment by picca** [comment]
Nice I did not know about this. When I build this in an chroot this is not a problem.
This is only when I build silx locally that I have this issue.
I think that I will force the binding like this to solve my issue.

thanks

**Comment by vasole** [comment]
The preferred binding is left to the application. 

At the ESRF BCU is still PyQt4. Taurus I do not know if they have finished the migration... BCU certainly not.

If our philosophy is not to force anything to the end user, PySide cannot be abandoned because it is the only binding that does not impose a license to the end user.

**Comment by t20100** [comment]
Yes the application can always choose the binding, but we need to provide a default.

Can we close this issue?

**Comment by picca** [comment]
Hello, this is just that even if the BCU does not migrate, Debian will, so the next release of Debian will not contain PyQt4. (we have less that 2 years to fix this :).
Do you have an Idea of the amount of work for the BCU in order to migrate there code to PyQT5 (before PyQT6 ;)

**Comment by vasole** [comment]
Honestly, I do not think they will migrate. I think they will develop new applications based on silx and/or taurus.

**Comment by picca** [comment]
That means, you will have to maintain the python-qt4.xxx packages internally and their build dependencies (qt4 c++ library, ...). Everythings will be remove at some point... good luck ;).

**Comment by vasole** [comment]
They (not us) are still python 2.x only. The paradox is that at a certain point those libraries (Qt, PyQt, ...) had to be compiled locally because the distributions were lagging behind. I guess they will keep running the old applications until the new ones are ready. If new development is based on silx and taurus I guess those 2 or 3 years will be enough. By other hand they are well advanced compared to us in what concerns web based interfaces.

**Comment by t20100** [comment]
We decided to change order and use PyQt5 first for v0.8.
Documentation and sx module need updates as IPython %pylab qt will no more be compatible with Qt

**Comment by t20100** [comment]
Closed by #1686
